### PR TITLE
Fix float equality and non-panicking decimal division

### LINKED ISSUE #364: [Bug]: Float and decimal division should never cause a panic
### Description

Relevant section of the [spec](https://ballerina.io/spec/lang/master/#section_6.23)

> float, decimal

> * performs the multiplication operation with the destination format being the same as the source format, as defined by IEEE 754-2008; this never causes a panic to occur
> / performs the division operation with the destination format being the same as the source format, as defined by IEEE 754-2008; this never causes a panic to occur
> % performs a remainder operation; the remainder is not the IEEE-defined remainder operation but is instead a remainder operation analogous to integer remainder; more precisely,
> if x is NaN or y is NaN or x is an infinity or y is a zero, then x % y is NaN
> for finite x, and infinite y, x % y is x
> for finite x and finite non-zero y, x % y is equal to the result of rounding x - (y × n) to the nearest representable value using the roundTiesToEven rounding mode, where n is the integer that is nearest to the mathematical quotient of x and y without exceeding it in magnitude; if the result is zero, then its sign is that of x
> no exceptions are thrown

### Steps to Reproduce

_No response_

### Affected Version(s)

_No response_

### OS, DB, other environment details and versions

_No response_

### Related area

Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### LINKED ISSUE #391: [Bug]: Float ==/!= semantics were wrong for NaN in runtime
### Description

$Subject

### Steps to Reproduce

_No response_

### Affected Version(s)

_No response_

### OS, DB, other environment details and versions

_No response_

### Related area

Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by warunalakshitha**
## Purpose

Fixes https://github.com/ballerina-platform/ballerina-lang-go/issues/364
Fixes https://github.com/ballerina-platform/ballerina-lang-go/issues/391

## Summary by CodeRabbit

* **Tests**
  * Added new test cases for NaN float value handling and edge case scenarios.
  * Expanded integration test fixtures for float, comparable, and const operations.
  * Updated test expectations for float remainder operations.

* **Bug Fixes**
  * Improved float equality comparison semantics for edge cases involving NaN and signed zero values.

[Review Change Stack](https://app.coderabbit.ai/change-stack/ballerina-platform/ballerina-lang-go/pull/458)

### CODE REVIEW

File: values/equal.go [hunk 1]
```diff
@@ -51,7 +51,7 @@ func deepEqual(v1, v2 BalValue, visited map[refPair]struct{}) bool {
 		return ok && a == b
 	case *decimal.Decimal:
 		b, ok := v2.(*decimal.Decimal)
-		return ok && a.Cmp(b) == 0
+		return ok && !a.IsNaN() && !b.IsNaN() && a.Cmp(b) == 0
```
Comment (heshanpadmasiri) [inline values/equal.go]: I don't think this is needed according to spec "+0 and -0 are not distinguished; if the coefficent is zero, then the sign is also constrained to be zero;
NaN, infinities and subnormals are not supported; operations that would result in one of these values according to the normal rules of IEEE 754-2008 instead result in a panic."

File: values/compare.go [hunk 2]
```diff
@@ -81,6 +81,9 @@ func Compare(x, y BalValue) CompareResult {
 		return CmpEQ
 	case *decimal.Decimal:
 		v2 := y.(*decimal.Decimal)
+		if v1.IsNaN() || v2.IsNaN() {
+			return CmpUN
+		}
```
Comment (heshanpadmasiri) [inline values/compare.go]: I don't think this can happen as well (Decimals can never be NaN in Ballerina)

File: decimal/decimal.go [hunk 3]
```diff
@@ -27,21 +27,34 @@ const (
...
+	c := newDecimalAPDContext()
 	c.Traps = apd.SystemOverflow | apd.SystemUnderflow | apd.Overflow |
 		apd.DivisionUndefined | apd.DivisionByZero | apd.DivisionImpossible |
 		apd.InvalidOperation
 	return c
 }()
 
+// ieeeContext matches context's precision and exponent range but has traps
+// disabled so IEEE arithmetic can yield NaN or infinity instead of *Error.
+var ieeeContext = func() *apd.Context {
```
Comment (heshanpadmasiri) [inline decimal/decimal.go]: Not sure these changes to decimal is needed.

File: values/equal.go (hunk 1 above)
Comment (warunalakshitha) [reply values/equal.go]: Removed decimal fixes

File: corpus/bal/subset8/08-float/25-v.bal [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// @productions float multiplicative-expr equality-expr floating-point-literal return-stmt function-call-expr assign-stmt local-var-decl-stmt
+import ballerina/io;
+
+public function main() {
```
Comment (warunalakshitha) [inline corpus/bal/subset8/08-float/25-v.bal:20]: This was added part of last PR comment

**Review by heshanpadmasiri — APPROVED** [review]
LGTM
